doc/developer/arrangements.md

```diff
@@ -0,0 +1,129 @@
+# Understanding what Costs in Materialize
+
+Materialize builds dataflow graphs that represents queries.
+The operators in these graphs can either be *stateless* (as in a Map, Filter, or Project operator) or *stateful*.
+Stateful operators hold on to arrangements of records, which are the main contributor to memory footprint in Materialize.
+Arrangements can be shared between operators that need the same data, which can ameliorate some of the cost.
+
+Let's talk through what an arrangement is, which operators require arrangements, and what we can do about them.
+
+## What is an "Arrangement"?
+
+An arrangement is an alternative representation of a stream of update triples `(data, time, diff)`.
+In stream form, the triple pass by and are gone.
+An arrangement records these triples, and indexes them by `data`.
+With some caveats about "logical compaction", the index maintains the *current* accumulation of `diff` for each `(data, time)`.
```

Comment:
Is there an URL that `logical compaction` can point to so that the reader understands what that is?

Reply:
Sorry, owing to an email UX issue I only saw the comment up above and so just landed it. Lemme clean up some of these issues in a follow-on.